cpp/oneapi/dal/backend/memory.hpp

```diff
@@ -80,6 +82,161 @@ class usm_deleter {
+                                      std::int64_t block_count,
+                                      std::int64_t src_stride_in_bytes,
+                                      std::int64_t block_size_in_bytes,
+                                      const event_vector& deps = {}) {
+    ONEDAL_ASSERT(dst_host);
+    ONEDAL_ASSERT(src_device);
+    ONEDAL_ASSERT(block_count > 0);
+    ONEDAL_ASSERT(src_stride_in_bytes > 0);
+    ONEDAL_ASSERT(block_size_in_bytes > 0);
+    ONEDAL_ASSERT(src_stride_in_bytes >= block_size_in_bytes);
+    ONEDAL_ASSERT(is_device_accessible_usm_pointer(q, src_device));
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::int64_t, block_count, block_size_in_bytes);
+
+    const auto gathered_device_unique =
+        make_unique_usm_device(q, block_count * block_size_in_bytes);
+
+    auto gather_event = q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+
+        const byte_t* src_byte = reinterpret_cast<const byte_t*>(src_device);
+        byte_t* gathered_byte = reinterpret_cast<byte_t*>(gathered_device_unique.get());
+
+        const std::int64_t local_size = std::min(down_pow2(block_count), 256L);
+        const auto range = make_multiple_nd_range_1d(block_count, local_size);
+
+        cgh.parallel_for(range, [=](sycl::nd_item<1> id) {
+            const auto i = id.get_global_id();
+            if (i < block_count) {
+                // TODO: Unroll for optimization
+                for (int j = 0; j < block_size_in_bytes; j++) {
+                    gathered_byte[i * block_size_in_bytes + j] =
+                        src_byte[i * src_stride_in_bytes + j];
+                }
+            }
+        });
+    });
+
+    auto copy_event = q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(gather_event);
+        cgh.memcpy(dst_host, gathered_device_unique.get(), block_count * block_size_in_bytes);
+    });
+
+    // We need to wait until gather kernel is completed to deallocate
+    // `gathered_device_unique`
+    copy_event.wait_and_throw();
+
+    return sycl::event{};
+}
+
+inline sycl::event scatter_host2device(sycl::queue& q,
+                                       void* dst_device,
+                                       const void* src_host,
+                                       std::int64_t block_count,
+                                       std::int64_t dst_stride_in_bytes,
+                                       std::int64_t block_size_in_bytes,
+                                       const event_vector& deps = {}) {
+    ONEDAL_ASSERT(dst_device);
+    ONEDAL_ASSERT(src_host);
+    ONEDAL_ASSERT(block_count > 0);
+    ONEDAL_ASSERT(dst_stride_in_bytes > 0);
+    ONEDAL_ASSERT(block_size_in_bytes > 0);
+    ONEDAL_ASSERT(dst_stride_in_bytes >= block_size_in_bytes);
+    ONEDAL_ASSERT(is_device_accessible_usm_pointer(q, dst_device));
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::int64_t, block_count, block_size_in_bytes);
+
+    const auto gathered_device_unique =
```

Comment:
No it's gathered, because we copy from the gathered host representation to scattered device, so `gathered_device_unique` is the copy of gathered host data on the device.